### PR TITLE
Proper memory management in Skwasm

### PR DESCRIPTION
**Pull request by eyebrowsoffire**
This fixes https://github.com/flutter/flutter/issues/127243.

This ensures that native skwasm objects are cleaned up when their associated dart-side objects are garbage collected.

### CODE REVIEW

File: lib/web_ui/lib/src/engine/skwasm/skwasm_impl/canvas.dart [hunk 1]
```diff
@@ -18,9 +18,9 @@ class SkwasmCanvas implements SceneCanvas {
   SkwasmCanvas.fromHandle(this._handle);
   CanvasHandle _handle;
 
-  void delete() {
-    canvasDestroy(_handle);
-  }
+  // Note that we do not need to deal with the finalizer registry here, because
+  // the underlying native skia object is tied directly to the lifetime of the
+  // associated SkPictureRecorder.
```
Comment (yjbanov) [inline lib/web_ui/lib/src/engine/skwasm/skwasm_impl/canvas.dart:23]: Good point. I wonder if in debug mode we should watch for recorders dropped on the floor and issue an error.

File: lib/web_ui/lib/src/engine/skwasm/skwasm_impl/filters.dart [hunk 2]
```diff
@@ -60,16 +62,20 @@ class SkwasmImageFilter implements SceneImageFilter {
...
-    }
-  }
+  static final SkwasmFinalizationRegistry<RawImageFilter> _registry =
+    SkwasmFinalizationRegistry<RawImageFilter>(imageFilterDispose);
 
+  @override
   final ImageFilterHandle handle;
   bool _isDisposed = false;
 
+  void dispose() {
```
Comment (kevmoo) [inline lib/web_ui/lib/src/engine/skwasm/skwasm_impl/filters.dart:72]: Drive-by-comment: this code is duplicated a LOT – worth making it a `mixin` or something?

File: lib/web_ui/lib/src/engine/skwasm/skwasm_impl/filters.dart [hunk 3]
```diff
@@ -100,33 +108,43 @@ class SkwasmColorFilter {
...
   ) => SkwasmColorFilter._(colorFilterCompose(outer.handle, inner.handle));
 
-  void dispose() {
-    if (!_isDisposed) {
-      colorFilterDispose(handle);
-      _isDisposed = true;
-    }
-  }
+  static final SkwasmFinalizationRegistry<RawColorFilter> _registry =
+    SkwasmFinalizationRegistry<RawColorFilter>(colorFilterDispose);
```
Comment (yjbanov) [inline lib/web_ui/lib/src/engine/skwasm/skwasm_impl/filters.dart:112]: Per-type finalization registry is an interesting choice. Why can't we use one registry for all types? I'm not sure what the trade-offs are actually.

File: lib/web_ui/lib/src/engine/skwasm/skwasm_impl/font_collection.dart [hunk 4]
```diff
@@ -19,28 +19,38 @@ import 'package:ui/src/engine/skwasm/skwasm_impl.dart';
 const String _robotoUrl =
     'https://fonts.gstatic.com/s/roboto/v20/KFOmCnqEu92Fr1Me5WZLCzYlKw.ttf';
 
-class SkwasmTypeface {
-  SkwasmTypeface(SkDataHandle data) : handle = typefaceCreate(data);
+class SkwasmTypeface implements SkwasmObjectWrapper<RawTypeface> {
```
Comment (yjbanov) [inline lib/web_ui/lib/src/engine/skwasm/skwasm_impl/font_collection.dart:22]: Do we even hold onto typefaces? Could we accidentally delete a typeface that's still in use?

File: lib/web_ui/lib/src/engine/skwasm/skwasm_impl/image.dart [hunk 5]
```diff
@@ -9,8 +9,10 @@ import 'package:ui/src/engine.dart';
 import 'package:ui/src/engine/skwasm/skwasm_impl.dart';
 import 'package:ui/ui.dart' as ui;
 
-class SkwasmImage implements ui.Image {
-  SkwasmImage(this.handle);
+class SkwasmImage implements SkwasmObjectWrapper<RawImage>, ui.Image {
+  SkwasmImage(this.handle) {
```
Comment (yjbanov) [inline lib/web_ui/lib/src/engine/skwasm/skwasm_impl/image.dart:13]: According to the `clone()` method below, it is the responsibility of the caller of the constructor to "pre-bump" the ref count of the handle. Does that hold for all invocations of this constructor?

File: lib/web_ui/lib/src/engine/skwasm/skwasm_impl/filters.dart (hunk 2 above)
Comment (eyebrowsoffire) [reply lib/web_ui/lib/src/engine/skwasm/skwasm_impl/filters.dart:72]: I took a couple different passes at it trying to DRY this out, this is actually more concise than it was in my previous passes. I did try actually factoring this out into a mixin, but there were two things that made it difficult:

(1) Each object has a different cleanup function for its handle, and since a mixin can't have a generative constructor it's hard for the adopting class to pass it in.
(2) The finalization registry is a static on the derived class, so the derived class can access it. I can't put the registry on the mixin because you'd only get one finalization registry for the mixin and not one per derived class.

There are probably clever ways to avoid these issues, but at a certain point I stopped refactoring and figured this is good enough. If you feel strongly, I can go back and try to figure out a way to fold some of this logic into `SkwasmObjectWrapper` though... maybe...

Comment (kevmoo) [reply lib/web_ui/lib/src/engine/skwasm/skwasm_impl/filters.dart:72]: Looking at the code, it's clear we'd need some funky new language features to generalize this. Nevermind. 😢

File: lib/web_ui/lib/src/engine/skwasm/skwasm_impl/memory.dart [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+
+import 'package:ui/src/engine.dart';
+
+abstract class SkwasmObjectWrapper<T extends NativeType> {
+  Pointer<T> get handle;
+}
+
+typedef DisposeFunction<T extends NativeType> = void Function(Pointer<T>);
+
+class SkwasmFinalizationRegistry<T extends NativeType> {
```
Comment (yjbanov) [inline lib/web_ui/lib/src/engine/skwasm/skwasm_impl/memory.dart:16]: Do we actually need the `<T>` type parameter? If not, then I think we can use a shared `DomFinalizationRegistry` instance for all objects. But, again, I'm not sure what kind of tradeoff exists in using one vs many registries 🤷  Just pointing out a possibility.

File: lib/web_ui/lib/src/engine/skwasm/skwasm_impl/paint.dart [hunk 7]
```diff
@@ -8,15 +8,26 @@ import 'package:ui/src/engine.dart';
...
+class SkwasmPaint implements SkwasmObjectWrapper<RawPaint>, ui.Paint {
+  factory SkwasmPaint() => SkwasmPaint._fromHandle(paintCreate());
+
+  SkwasmPaint._fromHandle(this.handle) {
+    _registry.register(this);
+  }
+
+  void dispose() {
+    assert(!_isDisposed);
+    _registry.unregister(this);
```
Comment (yjbanov) [inline lib/web_ui/lib/src/engine/skwasm/skwasm_impl/paint.dart:20]: This pattern is used in all types. I wonder it can be rolled into `SkwasmFinalizationRegistry`. We could have a nicely symmetric API:

```dart
registry.register(this);
registry.dispose(this); // also unregisters since it doesn't need to be GC'd any more
```

File: lib/web_ui/lib/src/engine/skwasm/skwasm_impl/memory.dart [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+  SkwasmFinalizationRegistry(DisposeFunction<T> dispose)
+    : registry = DomFinalizationRegistry(((JSNumber address) =>
+      dispose(Pointer<T>.fromAddress(address.toDart.toInt()))
+    ).toJS);
+
+  final DomFinalizationRegistry registry;
+
+  void register(SkwasmObjectWrapper<T> wrapper) {
```
Comment (yjbanov) [inline lib/web_ui/lib/src/engine/skwasm/skwasm_impl/memory.dart:24]: One nice feature of `UniqueRef`/`CountedRef` is that they integrate with `Instrumentation` which helps you find leaks and double-frees. But maybe this can be added in a follow-up PR, if the design allows it.

File: lib/web_ui/lib/src/engine/skwasm/skwasm_impl/filters.dart (hunk 3 above)
Comment (kevmoo) [reply lib/web_ui/lib/src/engine/skwasm/skwasm_impl/filters.dart:112]: Because each type has its own dispose function?

File: lib/web_ui/lib/src/engine/skwasm/skwasm_impl/filters.dart (hunk 2 above)
Comment (yjbanov) [reply lib/web_ui/lib/src/engine/skwasm/skwasm_impl/filters.dart:72]: `SkwasmFinalizationRegistry` already takes a reference to the type-specific clean-up function.  Maybe this is doable via `SkwasmFinalizationRegistry`. (see https://github.com/flutter/engine/pull/42328#discussion_r1206056640)

File: lib/web_ui/lib/src/engine/skwasm/skwasm_impl/filters.dart (hunk 3 above)
Comment (kevmoo) [reply lib/web_ui/lib/src/engine/skwasm/skwasm_impl/filters.dart:112]: Back to the mixin discussion, you could have a `Map<Type, DisposeFunction>`. Then you have one registry, but you need to do a map access on every dispose. 🤷

Comment (eyebrowsoffire) [reply lib/web_ui/lib/src/engine/skwasm/skwasm_impl/filters.dart:112]: Each handle has a different native disposal function. The alternative would be to have a single finalization registry, but add an extra layer of indirection that wraps the handle and is polymorphic on the dispose method (this is basically what you're doing on the CK side with your Ref objects). For the skwasm case, I found it a little cleaner to just do the per-object finalization registry and not end up with an extra layer of indirection.

File: lib/web_ui/lib/src/engine/skwasm/skwasm_impl/font_collection.dart (hunk 4 above)
Comment (eyebrowsoffire) [reply lib/web_ui/lib/src/engine/skwasm/skwasm_impl/font_collection.dart:22]: The typeface is a shared pointer under the hood, and the font collection hangs onto it. The `SkwasmFontCollection` does hold onto typeface objects explicitly. We used to need it in order to scan incoming text to find missing code points, but Skia does that for it for us now. However, when we reset the font collection (which is just a debug only situation for now), we re-register the base Roboto font with the new collection, so keeping the typeface around is useful in that scenario (rather than having to recreate it from data again). So I don't know if there is a super important reason to hang onto them in prod right now, but I don't think it hurts anything and makes the debug reset part easier.

Comment (eyebrowsoffire) [reply lib/web_ui/lib/src/engine/skwasm/skwasm_impl/font_collection.dart:22]: Rereading my comment, I'm not sure my first point was clear, but what I mean is that we don't have to worry about the typefaces dangling if we don't hang onto them, because the FontCollection has its own shared reference to each typeface. Our hanging onto it or not hanging onto it won't affect the operation of the FontCollection itself.

File: lib/web_ui/lib/src/engine/skwasm/skwasm_impl/image.dart (hunk 5 above)
Comment (eyebrowsoffire) [reply lib/web_ui/lib/src/engine/skwasm/skwasm_impl/image.dart:13]: Basically yes. This constructor (and all skwasm object wrappers constructors that just take a raw handle) expect a handle that has +1 ref count on it, and the object basically takes ownership of that ref count. All the native `create` functions return a handle with +1 ref count on it, and when cloning we bump the ref count for this same reason.

File: lib/web_ui/lib/src/engine/skwasm/skwasm_impl/memory.dart (hunk 6 above)
Comment (eyebrowsoffire) [reply lib/web_ui/lib/src/engine/skwasm/skwasm_impl/memory.dart:16]: We do need this type parameter, because each finalization registry has its own disposal function that it invokes and that disposal function takes a specific pointer type. I mentioned this in the other comment that it is possible to do this with another wrapper which basically "type-erases" this problem away, but I'd rather not add another layer of indirection.

File: lib/web_ui/lib/src/engine/skwasm/skwasm_impl/memory.dart [hunk 9]
```diff
@@ -7,25 +7,40 @@ import 'dart:js_interop';
...
 
 class SkwasmFinalizationRegistry<T extends NativeType> {
-  SkwasmFinalizationRegistry(DisposeFunction<T> dispose)
+  SkwasmFinalizationRegistry(this.dispose)
     : registry = DomFinalizationRegistry(((JSNumber address) =>
       dispose(Pointer<T>.fromAddress(address.toDart.toInt()))
     ).toJS);
 
   final DomFinalizationRegistry registry;
+  final DisposeFunction<T> dispose;
```
Comment (kevmoo) [inline lib/web_ui/lib/src/engine/skwasm/skwasm_impl/memory.dart:36]: nit: could these fields be private?

File: lib/web_ui/lib/src/engine/skwasm/skwasm_impl/memory.dart [hunk 10]
```diff
@@ -7,25 +7,40 @@ import 'dart:js_interop';
...
-  Pointer<T> get handle;
+class SkwasmObjectWrapper<T extends NativeType> {
+  SkwasmObjectWrapper(this.handle, this.registry) {
+    registry.register(this);
+  }
+  final SkwasmFinalizationRegistry<T> registry;
+  final Pointer<T> handle;
+  bool _isDisposed = false;
+
+  void dispose() {
```
Comment (kevmoo) [inline lib/web_ui/lib/src/engine/skwasm/skwasm_impl/memory.dart:18]: LOVE!

**Review by yjbanov — APPROVED** [review]
lgtm

File: lib/web_ui/lib/src/engine/skwasm/skwasm_impl/memory.dart (hunk 9 above)
Comment (yjbanov) [reply lib/web_ui/lib/src/engine/skwasm/skwasm_impl/memory.dart:36]: One unusual aspect of our codebase is that everything under `src/engine/` is automatically private, because we rewrite the code into `dart:_engine`. So it's OK to leave things public if it helps with testing and such.

File: lib/web_ui/lib/src/engine/skwasm/skwasm_impl/memory.dart (hunk 10 above)
Comment (yjbanov) [reply lib/web_ui/lib/src/engine/skwasm/skwasm_impl/memory.dart:18]: Aye!

File: lib/web_ui/lib/src/engine/skwasm/skwasm_impl/memory.dart (hunk 9 above)
Comment (kevmoo) [reply lib/web_ui/lib/src/engine/skwasm/skwasm_impl/memory.dart:36]: defaulting to private makes it easy to understand what's used. It also helps find when things are no longer used.
